File: forms/plastic-surgery-assessment/back-end-with-loco/src/engine/plastics_grader.rs
use super::flagged_issues::detect_additional_flags;
use super::plastics_rules::all_rules;
use super::types::{AssessmentData, FiredRule, GradingResult, RiskLevel};

/// Derive ASA class (1..=5) from clinician input.
fn derive_asa_class(data: &AssessmentData) -> Option<i32> {
    match data.anaesthetic_risk.asa_class.as_str() {
        "1" => Some(1),
        "2" => Some(2),
        "3" => Some(3),
        "4" => Some(4),
        "5" => Some(5),
        _ => None,
    }
}

/// Derive wound class (1..=4) from clinician input.
fn derive_wound_class(data: &AssessmentData) -> Option<i32> {
    match data.wound_tissue_assessment.wound_classification.as_str() {
        "clean" => Some(1),
        "clean-contaminated" => Some(2),
        "contaminated" => Some(3),
        "dirty" => Some(4),
        _ => None,
    }
}

/// Derive surgical complexity score (1..=4) from clinician input.
fn derive_complexity_score(data: &AssessmentData) -> Option<i32> {
    match data
        .procedure_planning_consent
        .procedure_complexity
        .as_str()
    {
        "1" => Some(1),
        "2" => Some(2),
        "3" => Some(3),
        "4" => Some(4),
        _ => None,
    }
}

/// Reduce fired rules + classifications to an overall risk level.
fn derive_overall_risk(
    fired_rules: &[FiredRule],
    asa_class: Option<i32>,
    wound_class: Option<i32>,
    complexity_score: Option<i32>,
) -> RiskLevel {
    let max_grade = fired_rules.iter().map(|r| r.grade).max().unwrap_or(0);

    if max_grade >= 4
        || asa_class.map(|v| v >= 4).unwrap_or(false)
        || wound_class == Some(4)
        || complexity_score == Some(4)
    {
        return "critical".to_string();
    }
    if max_grade >= 3
        || asa_class == Some(3)
        || wound_class == Some(3)
        || complexity_score == Some(3)
    {
        return "high".to_string();
    }
    if max_grade >= 2
        || asa_class == Some(2)
        || wound_class == Some(2)
        || complexity_score == Some(2)
    {
        return "moderate".to_string();
    }
    "low".to_string()
}

/// Evaluate every rule against the data, returning the rules that fired.
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            out.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                grade: rule.grade,
            });
        }
    }
    out
}

/// Pure function: grades a Plastic Surgery Assessment.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let fired_rules = run_rules(data);
    let additional_flags = detect_additional_flags(data);

    let asa_class = derive_asa_class(data);
    let wound_class = derive_wound_class(data);
    let complexity_score = derive_complexity_score(data);
    let overall_risk =
        derive_overall_risk(&fired_rules, asa_class, wound_class, complexity_score);

    GradingResult {
        asa_class,
        wound_class,
        complexity_score,
        overall_risk,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

File: forms/plastic-surgery-assessment/front-end-with-html/js/plastics-grader.js
// Pure plastic surgery grading engine. Mirrors the SvelteKit reference
// `plastics-grader.ts`: derives ASA class, wound class, complexity score,
// runs every declarative rule, and reduces fired-rule grades plus the three
// classifications down to an overall risk level.

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').FiredRule} FiredRule
 * @typedef {import('./types.js').GradingResult} GradingResult
 * @typedef {import('./types.js').RiskLevel} RiskLevel
 */

(function () {
'use strict';
window.PlasticSurgeryAssessment = window.PlasticSurgeryAssessment || {};
const { plasticsRules, detectAdditionalFlags } = window.PlasticSurgeryAssessment;

/** Derive ASA class from clinician input. */
function deriveASAClass(data) {
  const v = data.anaestheticRisk.asaClass;
  if (v === '1') return 1;
  if (v === '2') return 2;
  if (v === '3') return 3;
  if (v === '4') return 4;
  if (v === '5') return 5;
  return null;
}

/** Derive wound class from clinician input. */
function deriveWoundClass(data) {
  const v = data.woundTissueAssessment.woundClassification;
  if (v === 'clean') return 1;
  if (v === 'clean-contaminated') return 2;
  if (v === 'contaminated') return 3;
  if (v === 'dirty') return 4;
  return null;
}

/** Derive surgical complexity score from clinician input. */
function deriveComplexityScore(data) {
  const v = data.procedurePlanningConsent.procedureComplexity;
  if (v === '1') return 1;
  if (v === '2') return 2;
  if (v === '3') return 3;
  if (v === '4') return 4;
  return null;
}

/**
 * Reduce fired rules + classifications to an overall risk level.
 * @returns {RiskLevel}
 */
function deriveOverallRisk(firedRules, asaClass, woundClass, complexityScore) {
  const maxGrade = firedRules.length > 0
    ? Math.max(...firedRules.map((r) => r.grade))
    : 0;

  if (maxGrade >= 4 || (asaClass !== null && asaClass >= 4) || woundClass === 4 || complexityScore === 4) {
    return 'critical';
  }
  if (maxGrade >= 3 || asaClass === 3 || woundClass === 3 || complexityScore === 3) {
    return 'high';
  }
  if (maxGrade >= 2 || asaClass === 2 || woundClass === 2 || complexityScore === 2) {
    return 'moderate';
  }
  return 'low';
}

/**
 * Grade an assessment.
 * @param {AssessmentData} data
 * @returns {GradingResult}
 */
function calculatePlasticsGrade(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];

  for (const rule of plasticsRules) {
    try {
      if (rule.evaluate(data)) {
        firedRules.push({
          id: rule.id,
          category: rule.category,
          description: rule.description,
          grade: rule.grade
        });
      }
    } catch (e) {
      console.warn(`Plastics rule ${rule.id} evaluation failed:`, e);
    }
  }

  const asaClass = deriveASAClass(data);
  const woundClass = deriveWoundClass(data);
  const complexityScore = deriveComplexityScore(data);
  const overallRisk = deriveOverallRisk(firedRules, asaClass, woundClass, complexityScore);
  const additionalFlags = detectAdditionalFlags(data);

  return {
    asaClass,
    woundClass,
    complexityScore,
    overallRisk,
    firedRules,
    additionalFlags,
    timestamp: new Date().toISOString()
  };
}

window.PlasticSurgeryAssessment.calculatePlasticsGrade = calculatePlasticsGrade;
})();
